### PR TITLE
[RSN-9] - Implemented DTO models on frontend

### PR DESCRIPTION
**Pull request by wzarek**
## Description
Implemented models from DTO on backend to frontend with mappers from JSON string and object with unit tests.

## Related issue and/or ticket
RSN-9

### CODE REVIEW

File: Client/reasn-client/packages/common/models/EventDto.ts [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+        if (!('CreatedAt' in obj) || (typeof obj.CreatedAt !== 'string' && !(obj.CreatedAt instanceof Date))) throw new ModelMappingError('CommentDto','CreatedAt is required')
+        if (typeof obj.CreatedAt === 'string' && isNaN(Date.parse(obj.CreatedAt))) throw new ModelMappingError('CommentDto','CreatedAt is invalid')
+        if (!('UpdatedAt' in obj) || (typeof obj.UpdatedAt !== 'string' && !(obj.UpdatedAt instanceof Date))) throw new ModelMappingError('CommentDto','UpdatedAt is required')
+        if (typeof obj.UpdatedAt === 'string' && isNaN(Date.parse(obj.UpdatedAt))) throw new ModelMappingError('CommentDto','UpdatedAt is invalid')
+        if ('Slug' in obj === false || (typeof obj.Slug !== 'string')) throw new ModelMappingError('EventDto','Slug is required')
+        if ('StatusId' in obj === false || typeof obj.StatusId !== 'number') throw new ModelMappingError('EventDto','StatusId is required')
+        if ('Tags' in obj === false || obj.Tags === null || !Array.isArray(obj.Tags)) throw new ModelMappingError('EventDto','Tags is required')
```
Comment (raczu) [inline Client/reasn-client/packages/common/models/EventDto.ts:58]: The way to validate responses from API and creating the appropriate DTOs from them seems very complicated especially for DTOs with a large number of properties such as this one. What's more, the DTO by the definition should not have any methods so parsers or validators should be implemented as independent functions. I would consider making universal validator to somehow simplify this logic or use third party (lightweight) library to do so.

**Review by raczu — CHANGES_REQUESTED** [review]
After a long research and consideration of all possibilites, I think that the best solution will be to use [zod](https://zod.dev/) - it does not depend on other libraries and is lightweight. This will greatly simplify the verification of data received from the API. Additionally, we will be able to use this library also to verify the data provided by the user (data from forms).

See below example, which shows an easy way to handle date conversion to proper object:
Screenshot 2024-05-04 at 1 06 12 AM

**Review by raczu — CHANGES_REQUESTED** [review]
Apply those changes where its needed.

File: Client/reasn-client/packages/common/models/AddressDto.ts [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+    ZipCode: z.string().nullable()
+})
+
+export type AddressDto = z.infer<typeof AddressDtoSchema>
+
+export const AddressDtoMapper = {
+    fromObject: (entity: object): AddressDto => {
+        const result = AddressDtoSchema.safeParse(entity)
+        if (!result.success) {
+            throw new ModelMappingError('AddressDto', result.error.message)
```
[the remark this answers is not in the document]
Comment (wzarek) [reply Client/reasn-client/packages/common/models/AddressDto.ts:18]: Yes, correct - it is not a proper way of handling errors when using **parse**. But we don't use **parse** here :) I will expand our Mapping error to have also this validation error list, so it should be fine then.

File: Client/reasn-client/packages/common/models/CommentDto.ts [hunk 3]
```diff
@@ -0,0 +1,31 @@
+import ModelMappingError from '@reasn/common/errors/ModelMappingError'
+import { z } from "zod"
+
+export const CommentDtoSchema = z.object({
+    EventId: z.number(),
+    Content: z.string(),
+    CreatedAt: z.string().or(z.date()).transform(arg => new Date(arg)),
```
[the remark this answers is not in the document]
Comment (wzarek) [reply Client/reasn-client/packages/common/models/CommentDto.ts]: I have changed it a bit, I missed your comment there, so used zod string formats to check if it's a UTC formatted string or a Date type.

**Review by raczu — APPROVED** [review]

**Comment by wzarek** [comment]
@raczu if everything is done please merge, checked partially and looks fine

**Pull request merged**